lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -2,47 +2,107 @@
-      # set of hashes for memoized column conditions. Example:
-      #
-      #   {
-      #      Site => [
-      #        {name: 'a', city: 'b'},
-      #        {name: 'a', city: 'c'},
-      #        {name: 'b', city: 'b'},
-      #        {name: 'b', city: 'c'},
-      #      ],
-      #   }
-      #
-      @params ||= Hash.new do |models, model|
-        models[model] = []
-      end
+    def initialize(left = nil, right = nil, opt = nil)
+      @left = left
+      @right = right
+      @opt = opt
+      @plan = nil
     end
 
     def union(other)
       return unless other
 
-      # The tree is almost always right-heavy. Merge into the right node for better
-      # performance.
-      other.params.merge!(params) do |_, other_attrs_set, attrs_set|
-        if other_attrs_set.empty?
-          attrs_set
-        elsif attrs_set.empty?
-          other_attrs_set
-        else
-          attrs_set + other_attrs_set
+      self.class.new(self, other, __method__)
+    end
+
+    def product(other)
+      return unless other
+
+      self.class.new(self, other, __method__)
+    end
+
+    def should_cache?
+      plan!
+      return false if plan.dependency_size > RedisMemo::DefaultOptions.max_query_dependency_size
+
+      plan.model_attrs.each do |model, attrs_set|
+        return false if attrs_set.empty?
+
+        attrs_set.each do |attrs|
+          return false unless RedisMemo::MemoizeQuery
+            .memoized_columns(model)
+            .include?(attrs.keys.sort)
         end
       end
 
-      other
+      !plan.model_attrs.empty?
     end
 
-    def product(other)
+    def plan!
+      self.plan = Plan.new(self)
+      return if opt.nil?
+
+      left.plan!
+      right.plan!
+      __send__(:"plan_#{opt}")
+    end
+
+    #
+    # Extracted bind params is hash of sets: each key is a model class, each
+    # value is a set of hashes for memoized column conditions. Example:
+    #
+    #   {
+    #      Site => [
+    #        {name: 'a', city: 'b'},
+    #        {name: 'a', city: 'c'},
+    #        {name: 'b', city: 'b'},
+    #        {name: 'b', city: 'c'},
+    #      ],
+    #   }
+    #
+    def extract!
+      return if opt.nil?
+
+      left.extract!
+      right.extract!
+      __send__(:"#{opt}!")
+    end
+
+    def params
+      @params ||= Hash.new do |models, model|
+        models[model] = Set.new
+      end
+    end
+
+    protected
+
+    attr_accessor :left
+    attr_accessor :right
+    attr_accessor :opt
+    attr_accessor :plan
+
+    def union_attrs_set(left, right)
+      left.merge(right) do |_, attrs_set, other_attrs_set|
+        next attrs_set if other_attrs_set.empty?
+        next other_attrs_set if attrs_set.empty?
+
+        attrs_set + other_attrs_set
+      end
+    end
+
+    def plan_union
+      plan.dependency_size = left.plan.dependency_size + right.plan.dependency_size
+      plan.model_attrs = union_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
```

Comment:
Can we make `Plan` own the `union` method by passing in two `plan` obejcts?

Reply:
A `Plan` is a simple data hold object. It does not really makes sense for it to own `union`  since this method is used outside of the planning phase as well (see `union!`)